In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
df=pd.read_csv(r"C:\Users\kashr\Documents\Own machine learning Project\aw_fb_data.csv")

In [3]:
df.head()

Unnamed: 0  X1  age  gender  height  weight      steps  hear_rate  \
0           1   1   20       1   168.0    65.4  10.771429  78.531302   
1           2   2   20       1   168.0    65.4  11.475325  78.453390   
2           3   3   20       1   168.0    65.4  12.179221  78.540825   
3           4   4   20       1   168.0    65.4  12.883117  78.628260   
4           5   5   20       1   168.0    65.4  13.587013  78.715695   

    calories  distance  entropy_heart  entropy_setps  resting_heart  \
0   0.344533  0.008327       6.221612       6.116349           59.0   
1   3.287625  0.008896       6.221612       6.116349           59.0   
2   9.484000  0.009466       6.221612       6.116349           59.0   
3  10.154556  0.010035       6.221612       6.116349           59.0   
4  10.825111  0.010605       6.221612       6.116349           59.0   

   corr_heart_steps  norm_heart  intensity_karvonen  sd_norm_heart  \
0          1.000000   19.531302            0.138520       1.000000   
1          1.000000   19.453390            0.137967       1.000000   
2          1.000000   19.540825            0.138587       1.000000   
3          1.000000   19.628260            0.139208       1.000000   
4          0.982816   19.715695            0.139828       0.241567   

   steps_times_distance       device activity  
0              0.089692  apple watch    Lying  
1              0.102088  apple watch    Lying  
2              0.115287  apple watch    Lying  
3              0.129286  apple watch    Lying  
4              0.144088  apple watch    Lying

In [4]:
df.isnull().sum()


Unnamed: 0              0
X1                      0
age                     0
gender                  0
height                  0
weight                  0
steps                   0
hear_rate               0
calories                0
distance                0
entropy_heart           0
entropy_setps           0
resting_heart           0
corr_heart_steps        0
norm_heart              0
intensity_karvonen      0
sd_norm_heart           0
steps_times_distance    0
device                  0
activity                0
dtype: int64

In [5]:
df.duplicated().sum()


0

In [6]:
df.shape

(6264, 20)

In [7]:
df=df.rename(columns={df.columns[7]:'heart_rate'})

In [8]:
df=df.drop(columns=['X1','Unnamed: 0'])

In [9]:
df.tail()

age  gender  height  weight  steps  heart_rate  calories  distance  \
6259   46       0   157.5    71.4    1.0        35.0      20.5       1.0   
6260   46       0   157.5    71.4    1.0        35.0      20.5       1.0   
6261   46       0   157.5    71.4    1.0        35.0      20.5       1.0   
6262   46       0   157.5    71.4    1.0        35.0      20.5       1.0   
6263   46       0   157.5    71.4    1.0        35.0      20.5       1.0   

      entropy_heart  entropy_setps  resting_heart  corr_heart_steps  \
6259            0.0            0.0           35.0               1.0   
6260            0.0            0.0           35.0               1.0   
6261            0.0            0.0           35.0               1.0   
6262            0.0            0.0           35.0               1.0   
6263            0.0            0.0           35.0               1.0   

      norm_heart  intensity_karvonen  sd_norm_heart  steps_times_distance  \
6259         0.0                 0.0            0.0                   1.0   
6260         0.0                 0.0            1.0                   1.0   
6261         0.0                 0.0            1.0                   1.0   
6262         0.0                 0.0            1.0                   1.0   
6263         0.0                 0.0            1.0                   1.0   

      device        activity  
6259  fitbit  Running 7 METs  
6260  fitbit  Running 7 METs  
6261  fitbit  Running 7 METs  
6262  fitbit  Running 7 METs  
6263  fitbit  Running 7 METs

In [10]:
df['device'].value_counts()

device
apple watch    3656
fitbit         2608
Name: count, dtype: int64

In [11]:
df['device'].unique()

array(['apple watch', 'fitbit'], dtype=object)

In [12]:
df['activity'].unique()

array(['Lying', 'Sitting', 'Self Pace walk', 'Running 3 METs',
       'Running 5 METs', 'Running 7 METs'], dtype=object)

In [13]:
df.columns

Index(['age', 'gender', 'height', 'weight', 'steps', 'heart_rate', 'calories',
       'distance', 'entropy_heart', 'entropy_setps', 'resting_heart',
       'corr_heart_steps', 'norm_heart', 'intensity_karvonen', 'sd_norm_heart',
       'steps_times_distance', 'device', 'activity'],
      dtype='object')

In [14]:
df.sample()


age  gender  height  weight       steps  heart_rate  calories  distance  \
1581   31       1   191.0   105.7  595.908571        71.2  0.979604  0.461617   

      entropy_heart  entropy_setps  resting_heart  corr_heart_steps  \
1581       5.362901       5.787845           66.0          0.597206   

      norm_heart  intensity_karvonen  sd_norm_heart  steps_times_distance  \
1581         5.2            0.042276       4.418977            275.081374   

           device activity  
1581  apple watch  Sitting

In [15]:
df.columns

Index(['age', 'gender', 'height', 'weight', 'steps', 'heart_rate', 'calories',
       'distance', 'entropy_heart', 'entropy_setps', 'resting_heart',
       'corr_heart_steps', 'norm_heart', 'intensity_karvonen', 'sd_norm_heart',
       'steps_times_distance', 'device', 'activity'],
      dtype='object')

In [16]:
#  it is the library that I use For EDA purpose, because it gives the insights about the data in html file and it gives pretty good insights
from ydata_profiling import ProfileReport

In [17]:
prof=ProfileReport(df)
prof.to_file(output_file='output.html')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

In [18]:
# Feature Engineering

# Calculate BMI
df['BMI'] = df['weight'] / ((df['height'] / 100) ** 2)

# Define mental wellness conditions with adjusted thresholds
def determine_mental_wellness(row):
    score = 0
    if row['steps'] >= 3000:
        score += 1
    if row['resting_heart'] <= 85:
        score += 1
    if row['entropy_heart'] >= threshold_entropy:
        score += 1
    if row['calories'] >= 1500:
        score += 1
    if row['BMI'] <= 27:
        score += 1
    if 'Lying' not in row['activity'] and 'Sitting' not in row['activity']:
        score += 1
        
    return 1 if score >= 4 else 0  # Adjust threshold score as needed

# Set a threshold for entropy_heart based on analysis of your data
threshold_entropy = df['entropy_heart'].mean() * 0.8

# Apply the function to create the target column
df['mental_wellness'] = df.apply(determine_mental_wellness, axis=1)


In [19]:
df.head()

age  gender  height  weight      steps  heart_rate   calories  distance  \
0   20       1   168.0    65.4  10.771429   78.531302   0.344533  0.008327   
1   20       1   168.0    65.4  11.475325   78.453390   3.287625  0.008896   
2   20       1   168.0    65.4  12.179221   78.540825   9.484000  0.009466   
3   20       1   168.0    65.4  12.883117   78.628260  10.154556  0.010035   
4   20       1   168.0    65.4  13.587013   78.715695  10.825111  0.010605   

   entropy_heart  entropy_setps  resting_heart  corr_heart_steps  norm_heart  \
0       6.221612       6.116349           59.0          1.000000   19.531302   
1       6.221612       6.116349           59.0          1.000000   19.453390   
2       6.221612       6.116349           59.0          1.000000   19.540825   
3       6.221612       6.116349           59.0          1.000000   19.628260   
4       6.221612       6.116349           59.0          0.982816   19.715695   

   intensity_karvonen  sd_norm_heart  steps_times_distance       device  \
0            0.138520       1.000000              0.089692  apple watch   
1            0.137967       1.000000              0.102088  apple watch   
2            0.138587       1.000000              0.115287  apple watch   
3            0.139208       1.000000              0.129286  apple watch   
4            0.139828       0.241567              0.144088  apple watch   

  activity        BMI  mental_wellness  
0    Lying  23.171769                0  
1    Lying  23.171769                0  
2    Lying  23.171769                0  
3    Lying  23.171769                0  
4    Lying  23.171769                0

In [20]:
df['mental_wellness'].value_counts()

mental_wellness
0    3252
1    3012
Name: count, dtype: int64

In [21]:
df.columns

Index(['age', 'gender', 'height', 'weight', 'steps', 'heart_rate', 'calories',
       'distance', 'entropy_heart', 'entropy_setps', 'resting_heart',
       'corr_heart_steps', 'norm_heart', 'intensity_karvonen', 'sd_norm_heart',
       'steps_times_distance', 'device', 'activity', 'BMI', 'mental_wellness'],
      dtype='object')

In [23]:
df.describe()

age       gender       height       weight        steps  \
count  6264.000000  6264.000000  6264.000000  6264.000000  6264.000000   
mean     29.158525     0.476533   169.709052    69.614464   109.562268   
std       8.908978     0.499489    10.324698    13.451878   222.797908   
min      18.000000     0.000000   143.000000    43.000000     1.000000   
25%      23.000000     0.000000   160.000000    60.000000     5.159534   
50%      28.000000     0.000000   168.000000    68.000000    10.092029   
75%      33.000000     1.000000   180.000000    77.300000   105.847222   
max      56.000000     1.000000   191.000000   115.000000  1714.000000   

        heart_rate     calories     distance  entropy_heart  entropy_setps  \
count  6264.000000  6264.000000  6264.000000    6264.000000    6264.000000   
mean     86.142331    19.471823    13.832555       6.030314       5.739984   
std      28.648385    27.309765    45.941437       0.765574       1.256348   
min       2.222222     0.056269     0.000440       0.000000       0.000000   
25%      75.598079     0.735875     0.019135       6.108524       5.909440   
50%      77.267680     4.000000     0.181719       6.189825       6.157197   
75%      95.669118    20.500000    15.697188       6.247928       6.247928   
max     194.333333    97.500000   335.000000       6.475733       6.475733   

       resting_heart  corr_heart_steps   norm_heart  intensity_karvonen  \
count    6264.000000       6264.000000  6264.000000         6264.000000   
mean       65.869938          0.306447    20.272393            0.155479   
std        21.203017          0.775418    28.388116            0.210927   
min         3.000000         -1.000000   -76.000000           -2.714286   
25%        58.134333         -0.467303     1.148883            0.009819   
50%        75.000000          0.665829     9.820254            0.079529   
75%        76.138701          1.000000    27.077336            0.211868   
max       155.000000          1.000000   156.319444            1.297980   

       sd_norm_heart  steps_times_distance          BMI  mental_wellness  
count    6264.000000           6264.000000  6264.000000      6264.000000  
mean        8.110854            590.035239    24.042124         0.480843  
std        12.535080           4063.838530     3.271797         0.499673  
min         0.000000              0.000690    15.794307         0.000000  
25%         0.264722              0.659260    22.340297         0.000000  
50%         2.893503             13.368619    23.530366         0.000000  
75%         9.679672             93.728562    25.742187         1.000000  
max        74.457929          51520.000000    34.339634         1.000000

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6264 entries, 0 to 6263
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   age                   6264 non-null   int64  
 1   gender                6264 non-null   int64  
 2   height                6264 non-null   float64
 3   weight                6264 non-null   float64
 4   steps                 6264 non-null   float64
 5   heart_rate            6264 non-null   float64
 6   calories              6264 non-null   float64
 7   distance              6264 non-null   float64
 8   entropy_heart         6264 non-null   float64
 9   entropy_setps         6264 non-null   float64
 10  resting_heart         6264 non-null   float64
 11  corr_heart_steps      6264 non-null   float64
 12  norm_heart            6264 non-null   float64
 13  intensity_karvonen    6264 non-null   float64
 14  sd_norm_heart         6264 non-null   float64
 15  steps_times_distance 

In [25]:
# Encoding categorical values
df=pd.get_dummies(df,columns=['device','activity'],drop_first=True).astype(int)

In [26]:
df.sample()

age  gender  height  weight  steps  heart_rate  calories  distance  \
2331   29       0     163      61     47          90        19         0   

      entropy_heart  entropy_setps  ...  sd_norm_heart  steps_times_distance  \
2331              6              5  ...              3                     0   

      BMI  mental_wellness  device_fitbit  activity_Running 3 METs  \
2331   22                1              0                        0   

      activity_Running 5 METs  activity_Running 7 METs  \
2331                        0                        0   

      activity_Self Pace walk  activity_Sitting  
2331                        1                 0  

[1 rows x 24 columns]

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6264 entries, 0 to 6263
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   age                      6264 non-null   int32
 1   gender                   6264 non-null   int32
 2   height                   6264 non-null   int32
 3   weight                   6264 non-null   int32
 4   steps                    6264 non-null   int32
 5   heart_rate               6264 non-null   int32
 6   calories                 6264 non-null   int32
 7   distance                 6264 non-null   int32
 8   entropy_heart            6264 non-null   int32
 9   entropy_setps            6264 non-null   int32
 10  resting_heart            6264 non-null   int32
 11  corr_heart_steps         6264 non-null   int32
 12  norm_heart               6264 non-null   int32
 13  intensity_karvonen       6264 non-null   int32
 14  sd_norm_heart            6264 non-null   int32
 15  step

In [28]:
X=df.drop(columns='mental_wellness')
y=df['mental_wellness']

In [29]:
from sklearn.model_selection import train_test_split

In [30]:
X_train,X_test,y_train,y_test=train_test_split(X,y,random_state=42,test_size=0.2)

In [31]:
from sklearn.preprocessing import StandardScaler

# Standardize the features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [32]:
X_train_scaled

array([[-1.14900629,  1.05943653,  1.00753397, ...,  2.168132  ,
        -0.40663108, -0.41822491],
       [ 0.42674168, -0.94389798, -0.63364614, ..., -0.46122653,
        -0.40663108, -0.41822491],
       [-1.26155972,  1.05943653,  1.29715399, ...,  2.168132  ,
        -0.40663108, -0.41822491],
       ...,
       [-1.14900629,  1.05943653,  1.00753397, ..., -0.46122653,
        -0.40663108, -0.41822491],
       [-0.02347203, -0.94389798, -0.63364614, ...,  2.168132  ,
        -0.40663108, -0.41822491],
       [ 2.67781021, -0.94389798, -2.56444628, ..., -0.46122653,
        -0.40663108, -0.41822491]])

In [33]:
y_train

2024    1
3277    0
5015    1
2572    1
1024    1
       ..
3772    1
5191    0
5226    0
5390    1
860     1
Name: mental_wellness, Length: 5011, dtype: int32

In [34]:
X_test_scaled

array([[-1.03645287,  1.05943653,  1.393694  , ..., -0.46122653,
        -0.40663108,  2.39105796],
       [ 1.88993622, -0.94389798, -1.21288618, ..., -0.46122653,
        -0.40663108, -0.41822491],
       [-1.14900629,  1.05943653,  1.29715399, ..., -0.46122653,
        -0.40663108, -0.41822491],
       ...,
       [ 0.65184853, -0.94389798, -0.92326616, ..., -0.46122653,
        -0.40663108, -0.41822491],
       [-0.02347203, -0.94389798, -1.01980617, ...,  2.168132  ,
        -0.40663108, -0.41822491],
       [-0.69879259,  1.05943653,  1.10407398, ..., -0.46122653,
        -0.40663108,  2.39105796]])

In [35]:
from sklearn.linear_model import LogisticRegression

In [36]:
lr_model=LogisticRegression(penalty='l2')

In [37]:
lr_model.fit(X_train_scaled,y_train)

LogisticRegression()

In [38]:
y_pred=lr_model.predict(X_test_scaled)

In [39]:
from sklearn.metrics import accuracy_score

In [40]:
accuracy_score(y_test,y_pred)

0.9696727853152434

In [41]:
# Lets check the accuracy score on training data 

In [42]:
X_train_prediction=lr_model.predict(X_train_scaled)
training_data_accuracy=accuracy_score(X_train_prediction,y_train)
training_data_accuracy

0.9738575134703652

In [43]:
#  Lets check using random forest 

In [44]:
from sklearn.ensemble import RandomForestClassifier
rf_model=RandomForestClassifier(n_estimators=100,max_depth=None,max_samples=0.8,min_samples_leaf=1,min_samples_split=2,random_state=42)

In [45]:
rf_model.fit(X_train,y_train)

RandomForestClassifier(max_samples=0.8, random_state=42)

In [46]:
y_pred_rf=rf_model.predict(X_test)

In [47]:
accuracy_score(y_test,y_pred_rf)

1.0

In [48]:
from sklearn.model_selection import cross_val_score

In [49]:
cv_scores = cross_val_score(rf_model, X, y, cv=10, scoring='accuracy').mean()

In [50]:
cv_scores

0.9688564644256591

In [51]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [54]:
cv_scores = cross_val_score(lr_model, X_scaled, y, cv=10, scoring='accuracy').mean()

In [55]:
cv_scores

0.9420566009854727

In [63]:
from sklearn.neighbors import KNeighborsClassifier

In [64]:
knn_model=KNeighborsClassifier()

In [68]:
knn_model.fit(X_train_scaled,y_train)

KNeighborsClassifier()

In [69]:
knn_pred=knn_model.predict(X_test_scaled)

In [89]:
accuracy_score(y_test,knn_pred)

0.9936153232242618

In [90]:
cv_scores = cross_val_score(knn_model, X_scaled, y, cv=10, scoring='accuracy').mean()

In [91]:
cv_scores

0.8872978991189855

In [92]:
from xgboost import XGBClassifier

In [93]:
xgb_model= XGBClassifier(booster="gbtree",learning_rate=0.1,max_depth=6,min_child_weight=1,subsample=0.8,colsample_bytree=0.8,gamma=0,reg_lambda=1,reg_alpha=0,objective="binary:logistic",eval_metric="logloss")

In [94]:
xgb_model.fit(X_train_scaled,y_train)

XGBClassifier(base_score=None, booster='gbtree', callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=0.8, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric='logloss',
              feature_types=None, gamma=0, grow_policy=None,
              importance_type=None, interaction_constraints=None,
              learning_rate=0.1, max_bin=None, max_cat_threshold=None,
              max_cat_to_onehot=None, max_delta_step=None, max_depth=6,
              max_leaves=None, min_child_weight=1, missing=nan,
              monotone_constraints=None, multi_strategy=None, n_estimators=None,
              n_jobs=None, num_parallel_tree=None, random_state=None, ...)

In [95]:
xgb_pred=xgb_model.predict(X_test_scaled)

In [96]:
accuracy_score(y_test,xgb_pred)

1.0

In [97]:
cv_scores = cross_val_score(xgb_model, X_scaled, y, cv=10, scoring='accuracy').mean()

In [98]:
cv_scores

0.9862619808306707

# So After evaluating and applying cross validation(randomly splitting the dataset 10 times), Logistic gives accuracy of 94.21% and random forest gives 96.88% accuracy and knn gives 88.7% accuracy and XGboost is giving 98.6% accuracy

In [60]:
from sklearn.ensemble import VotingClassifier

# So we will select our best three model and apply voting ensemble to further improve model's overall performance

In [99]:
vc=VotingClassifier(estimators=[('rf_model',rf_model),('xgb_model',xgb_model),('lr_model',lr_model)])

In [100]:
vc.fit(X_train_scaled,y_train)

VotingClassifier(estimators=[('rf_model',
                              RandomForestClassifier(max_samples=0.8,
                                                     random_state=42)),
                             ('xgb_model',
                              XGBClassifier(base_score=None, booster='gbtree',
                                            callbacks=None,
                                            colsample_bylevel=None,
                                            colsample_bynode=None,
                                            colsample_bytree=0.8, device=None,
                                            early_stopping_rounds=None,
                                            enable_categorical=False,
                                            eval_metric='logloss',
                                            feature_types=None, gamma=...
                                            importance_type=None,
                                            interaction_constraints=None,
                                            learning_rate=0.1, max_bin=None,
                                            max_cat_threshold=None,
                                            max_cat_to_onehot=None,
                                            max_delta_step=None, max_depth=6,
                                            max_leaves=None, min_child_weight=1,
                                            missing=nan,
                                            monotone_constraints=None,
                                            multi_strategy=None,
                                            n_estimators=None, n_jobs=None,
                                            num_parallel_tree=None,
                                            random_state=None, ...)),
                             ('lr_model', LogisticRegression())])

In [101]:
vc_pred=vc.predict(X_test_scaled)

In [102]:
accuracy_score(y_test,vc_pred)

1.0

In [103]:
cv_scores = cross_val_score(vc, X_scaled, y, cv=10, scoring='accuracy').mean()

In [104]:
cv_scores

0.9832288752668775

# So at last after cross validation on each model and combing these model and also doing cross validation on combined model we can confidently say our model is giving 98.3% accuracy